keras_nlp/models/falcon/falcon_transformer_decoder.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from keras_nlp.backend import keras
+
+
+class FalconTransformerDecoder(keras.layers.Layer):
+    # TODO: set default values based on Falcon's config.
+    def __init__(
+        self,
+        num_attention_heads,
+        layer_norm_epsilon=1e-5,
+        attention_dropout=0,
+        feedforward_dropout=0,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.num_attention_heads = num_attention_heads
+        self.layer_norm_epsilon = layer_norm_epsilon
+        self.attention_dropout = attention_dropout
+        self.feedforward_dropout = feedforward_dropout
+
+    def build(self, decoder_sequence_shape):
+        self.hidden_dim = decoder_sequence_shape[-1]
+        self._input_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+        self._input_layernorm.build(decoder_sequence_shape)
+
+        # Attention layers.
+        # MHA for 1b model.
```

Comment:
What does this comment mean? Do we need a different attention mechanism for other sizes of models?

Reply:
Yes, I think different sizes have different attention mechanisms! 1b is MHA, 7b is MQA and 40b is something else!
Different attention mechanisms are handled [here](https://huggingface.co/tiiuae/falcon-rw-1b/blob/e4b9872bb803165eb22f0a867d4e6a64d34fce19/modeling_falcon.py#L211-L216).
Here are the configs for different sizes: [1b](https://huggingface.co/tiiuae/falcon-rw-1b/blob/e4b9872bb803165eb22f0a867d4e6a64d34fce19/config.json#L24-L25), [7b](https://huggingface.co/tiiuae/falcon-7b/blob/898df1396f35e447d5fe44e0a3ccaaaa69f30d36/config.json#L24-L25), [40b](https://huggingface.co/tiiuae/falcon-7b/blob/898df1396f35e447d5fe44e0a3ccaaaa69f30d36/config.json#L24-L25).

Reply:
As we discussed, to support different attention mechanism with one layer implementation, I will probably use Grouped Query Attention: https://github.com/keras-team/keras/blob/master/keras/layers/attention/grouped_query_attention.py